File: Embarcadero/XE8/ObjectPascal/Parallel Library/LifeUnitFmx.pas
unit LifeUnitFmx;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, FMX.Objects,
  FMX.Layouts, FMX.StdCtrls, LifeEngine;

const
  BoardSize: TSize = (cx: 500; cy: 500);

type
  TLifeForm = class(TForm)
    OpenDialog1: TOpenDialog;
    PaintBox1: TPaintBox;
    HorzScrollBar: TScrollBar;
    VertScrollBar: TScrollBar;
    Button1: TButton;
    Button2: TButton;
    Button3: TButton;
    Button4: TButton;
    CheckBox1: TCheckBox;
    Label1: TLabel;
    Label2: TLabel;
    Label3: TLabel;
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure Button1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure FormResize(Sender: TObject);
    procedure PaintBox1MouseDown(Sender: TObject; Button: TMouseButton;
      Shift: TShiftState; X, Y: Single);
    procedure PaintBox1MouseMove(Sender: TObject; Shift: TShiftState; X,
      Y: Single);
    procedure PaintBox1Paint(Sender: TObject; Canvas: TCanvas);
    procedure VertScrollBarChange(Sender: TObject);
    procedure HorzScrollBarChange(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button4Click(Sender: TObject);
    procedure Button3Click(Sender: TObject);
    procedure CheckBox1Change(Sender: TObject);
  private
    var
      FLifeEngine: TLifeEngine;
      FLifeBoard: TLifeBoard;
      FGensPerSecond, FMaxGensPerSecond: Double;
      FViewOffset, FViewSize: TPoint;
    procedure LifeEngineUpdate(Sender: TObject);
    { Private declarations }
  public
    { Public declarations }
  end;

var
  LifeForm: TLifeForm;

implementation

uses
  System.Math;

{$R *.fmx}

{ TLifeForm }

procedure TLifeForm.Button1Click(Sender: TObject);
begin
  if not FLifeEngine.Running then
  begin
    FLifeEngine.Start;
    Button1.Text := 'Stop';
  end else
  begin
    FLifeEngine.Stop;
    Button1.Text := 'Start';
  end;
end;

procedure TLifeForm.Button2Click(Sender: TObject);
begin
  if not FLifeEngine.Running then
  begin
    FLifeEngine.Clear;
    FLifeBoard := FLifeEngine.LifeBoard;
    FormResize(Sender);
    PaintBox1.InvalidateRect(PaintBox1.BoundsRect);
  end;
end;

procedure TLifeForm.Button3Click(Sender: TObject);
begin
  if not FLifeEngine.Running and OpenDialog1.Execute then
  begin
    FLifeEngine.LoadPattern(OpenDialog1.FileName);
  end;
end;

procedure TLifeForm.Button4Click(Sender: TObject);
begin
  HorzScrollBar.Value := (Length(FLifeBoard) - FViewSize.X) div 2;
  VertScrollBar.Value := (Length(FLifeBoard[0]) - FViewSize.Y) div 2;
end;

procedure TLifeForm.CheckBox1Change(Sender: TObject);
begin
  if FLifeEngine <> nil then
    FLifeEngine.Parallel := CheckBox1.IsChecked;
end;

procedure TLifeForm.FormCloseQuery(Sender: TObject; var CanClose: Boolean);
begin
  if FLifeEngine.Running then
    Button1Click(Sender);
  CanClose := not FLifeEngine.Running;
end;

procedure TLifeForm.FormCreate(Sender: TObject);
begin
  FLifeEngine := TLifeEngine.Create(BoardSize);
  FLifeEngine.OnUpdate := LifeEngineUpdate;
  FLifeBoard := FLifeEngine.LifeBoard;
  FLifeEngine.UpdateRate := 30;
end;

procedure TLifeForm.FormDestroy(Sender: TObject);
begin
  FLifeEngine.Free;
end;

procedure TLifeForm.FormResize(Sender: TObject);
begin
  FViewSize := Point((Trunc(PaintBox1.Width) - 10) div 10, (Trunc(PaintBox1.Height) - 10) div 10);
  HorzScrollBar.Max := Length(FLifeBoard){ - FViewSize.X};
  VertScrollBar.Max := Length(FLifeBoard[0]){ - FViewSize.Y};
end;

procedure TLifeForm.HorzScrollBarChange(Sender: TObject);
begin
  FViewOffset.X := Trunc(HorzScrollBar.Value);
  PaintBox1.InvalidateRect(PaintBox1.BoundsRect);
end;

procedure TLifeForm.LifeEngineUpdate(Sender: TObject);
begin
  FLifeBoard := FLifeEngine.LifeBoard;
  FGensPerSecond := FLifeEngine.GensPerSecond;
  FMaxGensPerSecond := FLifeEngine.MaxGensPerSecond;
  Label2.Text := Format('%f Generations Per Second', [FGensPerSecond]);
  Label3.Text := Format('%f Max Generations Per Second', [FMaxGensPerSecond]);
  PaintBox1.InvalidateRect(PaintBox1.BoundsRect);
end;

procedure TLifeForm.PaintBox1MouseDown(Sender: TObject; Button: TMouseButton;
  Shift: TShiftState; X, Y: Single);
var
  Row, Column: Integer;
begin
  if not FLifeEngine.Running and (Button = TMouseButton.mbLeft) then
  begin
    Row := Trunc(Y) div 10;
    Column := Trunc(X) div 10;
    if (Row >= 0) and (Row <= FViewSize.Y) and
      (Column >= 0) and (Column <= FViewSize.X) then
    begin
      FLifeBoard[FViewOffset.X + Column, FViewOffset.Y + Row] :=
        FLifeBoard[FViewOffset.X + Column, FViewOffset.Y + Row] xor 1;
        PaintBox1.InvalidateRect(TRectF.Create(Column + 10, Row * 10, Column * 10 + 11, Row * 10 + 11));
    end;
    Label1.Text := Format('%d, %d', [FViewOffset.X + Column, FViewOffset.Y + Row]);
  end;
end;

procedure TLifeForm.PaintBox1MouseMove(Sender: TObject; Shift: TShiftState; X,
  Y: Single);
begin
    Label1.Text := Format('%d, %d', [FViewOffset.X + Trunc(X) div 10,
      FViewOffset.Y + Trunc(Y) div 10]);
end;

procedure TLifeForm.PaintBox1Paint(Sender: TObject; Canvas: TCanvas);
var
  I, J: Integer;
  Gens, Max: Integer;
  ScaleTemp: Integer;
begin
  PaintBox1.Canvas.BeginScene();
  PaintBox1.Canvas.Stroke.Kind := TBrushKind.Solid;
  PaintBox1.Canvas.Stroke.Color := TAlphaColorRec.Gray;
  if Length(FLifeBoard) > 0 then
    for I := 0 to FViewSize.X - 1 do
      for J := 0 to FViewSize.Y - 1 do
      begin
        if FLifeBoard[Min(FViewOffset.X + I, High(FLifeBoard)), Min(FViewOffset.Y + J, High(FLifeBoard[0]))] <> 0 then
          with PaintBox1.Canvas do
          begin
            Fill.Color := TAlphaColorRec.Black;
            FillREct(TRectF.Create(I * 10, J * 10, I * 10 + 11, J * 10 + 11), 0, 0, [],100);
          end else
          with PaintBox1.Canvas do
          begin
            Fill.Color := Self.Canvas.Fill.Color;
            DrawRect(TRectF.Create(I * 10, J * 10, I * 10 + 11, J * 10 + 11), 0, 0, [], 100);
          end;
      end;
  with PaintBox1.Canvas do
  begin
    // Calculate the scale for the graph
    ScaleTemp := 1000000;
    while ScaleTemp > 10 do
    begin
      if FMaxGensPerSecond * 10 < ScaleTemp then
        ScaleTemp := ScaleTemp div 10
      else
        break;
    end;
    Gens := Trunc((FGensPerSecond * PaintBox1.Height) / ScaleTemp);
    Max := Trunc((FMaxGensPerSecond * PaintBox1.Height) / ScaleTemp);
    Fill.Color := TAlphaColorRec.Green;
    FillRect(TRectF.Create(PaintBox1.Width - 4, PaintBox1.Height - Gens, PaintBox1.Width,
      PaintBox1.Height), 0,0,[], 100);

    Stroke.Color := TAlphaColorRec.Red;
    DrawLine(TPointF.Create(PaintBox1.Width-4, PaintBox1.Height-Max),
    TPointF.Create(PaintBox1.Width, PaintBox1.Height - Max), 100);
  end;
  PaintBox1.Canvas.EndScene;
end;

procedure TLifeForm.VertScrollBarChange(Sender: TObject);
begin
  FViewOffset.Y := Trunc(VertScrollBar.Value);
//  PaintBox1.Repaint();
  PaintBox1.InvalidateRect(PaintBox1.BoundsRect);
end;

end.

File: Embarcadero/Seattle/CPP/Parallel Library/FMX/LifeUnitFmx.cpp
//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "LifeUnitFmx.h"
#include <cmath>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TLifeForm *LifeForm;
///---------------------------------------------------------------------------
__fastcall TLifeForm::TLifeForm(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
int __fastcall TLifeForm::MulDiv(int val, int num, int denom)
{
    	long retval = val;
	retval *= num;
	retval /= denom;
	return (int) retval;
}
//---------------------------------------------------------------------------
void __fastcall TLifeForm::Button1Click(TObject *Sender)
{
	if(!FLifeEngine->Running) {
		FLifeEngine->Start();
		Button1->Text = "&Stop";
	}
	else {
		FLifeEngine->Stop();
		Button1->Text = "&Start";
    }
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::Button2Click(TObject *Sender)
{
	if(!FLifeEngine->Running) {
		FLifeEngine->Clear();
		FLifeBoard = FLifeEngine->LifeBoard;
		FormResize(Sender);
		PaintBox1->InvalidateRect(PaintBox1->BoundsRect);
	}
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::FormCloseQuery(TObject *Sender, bool &CanClose)
{
	if(FLifeEngine->Running) {
		Button1Click(Sender);
	}
  	CanClose = !FLifeEngine->Running;
}
//---------------------------------------------------------------------------
void __fastcall TLifeForm::LifeEngineUpdate(TObject * Sender)
{
	FLifeBoard = FLifeEngine->LifeBoard;
	FGensPerSecond = FLifeEngine->GensPerSecond;
	FMaxGensPerSecond = FLifeEngine->MaxGensPerSecond;
	Label2->Text = Format("%f Generations Per Second", ARRAYOFCONST((FGensPerSecond)));
	Label3->Text = Format("%f Max Generations Per Second", ARRAYOFCONST((FMaxGensPerSecond)));
	PaintBox1->InvalidateRect(PaintBox1->BoundsRect);
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::FormCreate(TObject *Sender)
{
	BoardSize = TSize(500,500);
	FLifeEngine = new TLifeEngine(BoardSize);
	FLifeEngine->OnUpdate = LifeEngineUpdate;
	FLifeBoard = FLifeEngine->LifeBoard;
	FLifeEngine->UpdateRate = 30;
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::FormDestroy(TObject *Sender)
{
	FLifeEngine->Free();
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::FormResize(TObject *Sender)
{
	FViewSize = System::Classes::Point((PaintBox1->Width - 10) / 10, (PaintBox1->Height - 10) / 10);
	HorzScrollBar->Max = FLifeBoard.Length;
	VertScrollBar->Max = FLifeBoard[0].Length;
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::Button4Click(TObject *Sender)
{
	HorzScrollBar->Value = (FLifeBoard.Length - FViewSize.X) / 2;
	VertScrollBar->Value = (FLifeBoard[0].Length - FViewSize.Y) / 2;
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::Button3Click(TObject *Sender)
{
	if((!FLifeEngine->Running) && (OpenDialog1->Execute()))
	{
		FLifeEngine->LoadPattern(OpenDialog1->FileName);
	}
}
//---------------------------------------------------------------------------
void __fastcall TLifeForm::HorzScrollBarChange(TObject *Sender)
{
	FViewOffset.X = HorzScrollBar->Value;
	PaintBox1->InvalidateRect(PaintBox1->BoundsRect);
}
//---------------------------------------------------------------------------
void __fastcall TLifeForm::VertScrollBarChange(TObject *Sender)
{
	FViewOffset.Y = VertScrollBar->Value;
	PaintBox1->InvalidateRect(PaintBox1->BoundsRect);
}
//---------------------------------------------------------------------------
void __fastcall TLifeForm::PaintBox1MouseDown(TObject *Sender, TMouseButton Button,
		  TShiftState Shift, float X, float Y)
{
	if((!FLifeEngine->Running) && (Button == TMouseButton::mbLeft)) {
		int Row = Y / 10;
		int Column = X / 10;
		if((Row >= 0) && (Row <= FViewSize.Y) && (Column >= 0) && (Column <= FViewSize.X)) {
			FLifeBoard[FViewOffset.X + Column][FViewOffset.Y + Row] =
				FLifeBoard[FViewOffset.X + Column][FViewOffset.Y + Row] ^ 1;
			PaintBox1->InvalidateRect(TRectF(Column + 10, Row * 10, Column * 10 + 11, Row * 10 + 11));
		}
		Label1->Text = String().sprintf(L"%d, %d", FViewSize.X + Column, FViewSize.Y + Row);
	}
}
//---------------------------------------------------------------------------
void __fastcall TLifeForm::PaintBox1Paint(TObject *Sender, TCanvas *Canvas)
{
	PaintBox1->Canvas->BeginScene();
    PaintBox1->Canvas->Stroke->Kind = TBrushKind::Solid;
	PaintBox1->Canvas->Stroke->Color = static_cast<TAlphaColor>(TAlphaColorRec::Gray);
	if(FLifeBoard.Length > 0) {
		for(int i = 0; i < FViewSize.X; i++) {
			for(int j = 0; j < FViewSize.Y; j++) {
				if(FLifeBoard[Min(FViewOffset.X + i, FLifeBoard.High)][Min(FViewOffset.Y + j, FLifeBoard[0].High)] != 0)
				{

					PaintBox1->Canvas->Fill->Color = static_cast<TAlphaColor>(TAlphaColorRec::Black);
					PaintBox1->Canvas->FillRect(TRectF(i * 10, j * 10, i * 10 + 11, j * 10 + 11), 0, 0, TCorners(), 100);
				}
				else {
					PaintBox1->Canvas->Fill->Color = this->Canvas->Fill->Color;
					PaintBox1->Canvas->DrawRect(TRectF(i * 10, j * 10, i * 10 + 11, j * 10 + 11), 0, 0, TCorners(), 100);
				}
			}
		}
	}
	int Scale = 1000000;
	while(Scale > 10) {
		if((FMaxGensPerSecond * 10) < Scale) {
			Scale = Scale / 10;
		}
		else  {
			break;
		}
		int Gens = this->MulDiv(FGensPerSecond, PaintBox1->Height, Scale);
		int Max =  this->MulDiv(FMaxGensPerSecond, PaintBox1->Height, Scale);
		PaintBox1->Canvas->Fill->Color = static_cast<TAlphaColor>(TAlphaColorRec::Green);
		PaintBox1->Canvas->FillRect(TRectF(PaintBox1->Width - 4, PaintBox1->Height - Gens, PaintBox1->Width, PaintBox1->Height),
			0,0, TCorners(), 100);
		PaintBox1->Canvas->Stroke->Color = static_cast<TAlphaColor>(TAlphaColorRec::Red);
		PaintBox1->Canvas->DrawLine(TPointF(PaintBox1->Width - 4, PaintBox1->Height - Max),
			TPointF(PaintBox1->Width, PaintBox1->Height - Max), 100);
	}
	PaintBox1->Canvas->EndScene();
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::PaintBox1MouseMove(TObject *Sender, TShiftState Shift,
          float X, float Y)
{
	Label1->Text = String().sprintf(L"%d, %d", static_cast<int>(FViewOffset.X + X / 10),
		static_cast<int>(FViewOffset.Y + Y / 10));
}
//---------------------------------------------------------------------------

void __fastcall TLifeForm::CheckBox1Change(TObject *Sender)
{
	if(FLifeEngine != NULL) {
		FLifeEngine->Parallel = CheckBox1->IsChecked;
	}
}
//---------------------------------------------------------------------------

